File: unity/Assets/Scripts/Multiplayer/Rooms/Updates/PlayerUpdate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;
using static Michitai.Multiplayer.Time.Time;

namespace Michitai.Multiplayer.Rooms.Updates
{
    /// <summary>
    /// Information about a player update received in a room.
    /// Contains update details, sender information, and data.
    /// </summary>
    [System.Serializable]
    public class PlayerUpdate
    {
        [SerializeField]
        private string created_at;

        /// <summary>
        /// The unique ID of the update.
        /// </summary>
        public string update_id;

        /// <summary>
        /// The ID of the player who sent the update.
        /// </summary>
        public int from_player_id;

        /// <summary>
        /// The type of update.
        /// </summary>
        public string type;

        /// <summary>
        /// The update data as JSON string.
        /// </summary>
        public string data_json;    // Unity mode

        /// <summary>
        /// Timestamp when the update was created.
        /// </summary>
        public DateTimeOffset? CreatedAt
        {
            get
            {
                return ParseUtc(created_at);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/updates/PlayerUpdate.java
package com.michitai.multiplayer.rooms.updates;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Information about a player update received in a room.
 * Contains update details, sender information, and data.
 *
 * @param <T> The type to deserialize update data into.
 */
public class PlayerUpdate<T> {
    @JsonProperty("update_id")
    private String updateId;

    @JsonProperty("from_player_id")
    private int fromPlayerId;

    @JsonProperty("type")
    private String type;

    @JsonProperty("data_json")
    private String dataJson;

    @JsonProperty("data")
    private T data;

    @JsonProperty("created_at")
    private String createdAt;

    public String getUpdateId() {
        return updateId;
    }

    public void setUpdateId(String updateId) {
        this.updateId = updateId;
    }

    public int getFromPlayerId() {
        return fromPlayerId;
    }

    public void setFromPlayerId(int fromPlayerId) {
        this.fromPlayerId = fromPlayerId;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getDataJson() {
        return dataJson;
    }

    public void setDataJson(String dataJson) {
        this.dataJson = dataJson;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }
}
